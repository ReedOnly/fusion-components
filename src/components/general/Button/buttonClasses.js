import classNames from "classnames";
import { componentDisplayTypes } from "../../contexts/ComponentDisplayContext";
import styles from "./styles/index.less";

const getLayoutClasses = ({ block, flex }) => ({
    [styles.block]: block,
    [styles.flex]: flex,
});

const getMouseDownClasses = ({ mouseHasBeenDown }) => ({
    [styles.mouseHasBeenDown]: mouseHasBeenDown,
});

const getButtonVariantClasses = ({ contained, outlined, frameless }) => ({
    [styles.contained]: contained || (!outlined && !frameless), // Default to contained
    [styles.outlined]: outlined,
    [styles.frameless]: frameless,
    [styles.icon]: false,
});

const getButtonStyleClasses = ({ primary, signal }) => ({
    [styles.primary]: primary,
    [styles.signal]: signal,
});

const getButtonSizeClasses = ({ compact, comfortable, displayType }) => ({
    [styles.compact]: displayType === componentDisplayTypes.compact || compact,
    [styles.comfortable]:
        (displayType === componentDisplayTypes.comfortable &&
            displayType !== componentDisplayTypes.comfortable) ||
        (comfortable && !compact),
});

export default props =>
    classNames(
        props.className,
        styles.container,
        getLayoutClasses(props),
        getMouseDownClasses(props),
        getButtonVariantClasses(props),
        getButtonStyleClasses(props),
        getButtonSizeClasses(props)
    );
