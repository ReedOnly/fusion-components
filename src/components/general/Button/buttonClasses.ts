import classNames from 'classnames';
import { componentDisplayTypes } from '../../contexts/componentDisplayContext';
import styles from './styles/index.less';

const getLayoutClasses = ({ block, flex }) => ({
    [styles.block]: block,
    [styles.flex]: flex,
});

const getMouseDownClasses = ({ mouseHasBeenDown }) => ({
    [styles.mouseHasBeenDown]: mouseHasBeenDown,
});

const getButtonVariantClasses = ({ contained, outlined, frameless }) => ({
    [styles.contained]: contained || (!outlined && !frameless), // Default to contained
    [styles.outlined]: outlined,
    [styles.frameless]: frameless,
});

const getButtonStyleClasses = ({ primary, signal }) => ({
    [styles.primary]: primary,
    [styles.signal]: signal,
});

const getButtonSizeClasses = ({ compact, comfortable, displayType }) => ({
    [styles.compact]: displayType === componentDisplayTypes.compact || compact,
    [styles.comfortable]:
        (displayType === componentDisplayTypes &&
            displayType !== componentDisplayTypes.comfortable) ||
        (comfortable && !compact),
});

export default props =>
    classNames(
        props.className,
        styles.container,
        getLayoutClasses(props),
        getMouseDownClasses(props),
        getButtonVariantClasses(props),
        getButtonStyleClasses(props),
        getButtonSizeClasses(props)
    );
