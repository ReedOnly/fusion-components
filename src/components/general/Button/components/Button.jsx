import React, { forwardRef } from "react";
import Content from "./Content";
import getButtonClasses from "../buttonClasses";

const ButtonComponent = forwardRef(
    (
        {
            children,
            disabled,
            onMouseDown,
            onMouseUp,
            onClick,
            onClickCapture,
            ...props
        },
        ref
    ) => (
        <button
            type="button"
            className={getButtonClasses(props)}
            disabled={disabled}
            onMouseDown={onMouseDown}
            onMouseUp={onMouseUp}
            onClick={onClick}
            onClickCapture={onClickCapture}
            ref={ref}
        >
            <Content>{children}</Content>
        </button>
    )
);

ButtonComponent.displayName =
    "@fusion/components/general/Button/ButtonComponent";

export default ButtonComponent;
