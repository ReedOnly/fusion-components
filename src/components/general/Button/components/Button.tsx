import * as React from "react";
import Content from "./Content";
import getButtonClasses from "../buttonClasses";

type ButtonProps = {
    disabled?: boolean,
    onClick?: React.EventHandler<React.SyntheticEvent>,
    onClickCapture?: React.EventHandler<React.SyntheticEvent>,
    onMouseDown?: React.EventHandler<React.SyntheticEvent>,
    onMouseUp?: React.EventHandler<React.SyntheticEvent>,
    displayType: string,
    mouseHasBeenDown: boolean,
};

const ButtonComponent = React.forwardRef<HTMLButtonElement | null, React.PropsWithChildren<ButtonProps>>(
    ({ children, disabled, onMouseDown, onMouseUp, onClick, onClickCapture, ...props }, ref) => (
        <button
            type="button"
            className={getButtonClasses(props)}
            disabled={disabled}
            onMouseDown={onMouseDown}
            onMouseUp={onMouseUp}
            onClick={onClick}
            onClickCapture={onClickCapture}
            ref={ref}
        >
            <Content>{children}</Content>
        </button>
    )
);

ButtonComponent.defaultProps = {
    disabled: true,
    onClick: () => {},
    onClickCapture: () => {},
    onMouseDown: () => {},
    onMouseUp: () => {},
};

ButtonComponent.displayName = "@fusion/components/general/Button/ButtonComponent";

export default ButtonComponent;
