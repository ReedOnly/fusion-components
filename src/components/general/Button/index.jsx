import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import ComponentDisplayContext, {
    componentDisplayTypes,
} from "../../contexts/ComponentDisplayContext";
import styles from "./styles.less";

class Button extends Component {
    static propTypes = {
        children: PropTypes.node.isRequired,
        /** Disable the button */
        disabled: PropTypes.bool,

        // Variants
        /** Use the contained variant */
        contained: PropTypes.bool,
        /** Use the outlined variant */
        outlined: PropTypes.bool,
        /** Use the frameless variant */
        frameless: PropTypes.bool,

        // Styles
        /** Use the primary coloring */
        primary: PropTypes.bool,
        /** Use the signal coloring */
        signal: PropTypes.bool,

        // Sizes
        /** Small button */
        small: PropTypes.bool,
        /** Medium button */
        medium: PropTypes.bool,
        /** Large button */
        large: PropTypes.bool,

        // Link
        /** Provide an url and the button wil use <a>-tag instead of <button>-tag */
        url: PropTypes.string,
        /** Set target="_blank" */
        targetBlank: PropTypes.bool,
        onMouseDown: PropTypes.func,
        onClickCapture: PropTypes.func,
        onClick: PropTypes.func,
    };

    static defaultProps = {
        disabled: false,
        contained: false,
        outlined: false,
        frameless: false,
        primary: true,
        signal: false,
        small: true,
        medium: true,
        large: false,
        url: null,
        targetBlank: false,
        onMouseDown: () => {},
        onClickCapture: () => {},
        onClick: () => {},
    };

    state = {
        isMouseDown: false,
        mouseHasBeenDown: false,
    };

    onMouseUp() {
        this.setState({
            isMouseDown: false,
            mouseHasBeenDown: true,
        });

        if (this.button) {
            this.button.blur();
        }
    }

    onMouseDown(e) {
        const { onMouseDown } = this.props;
        if (onMouseDown) {
            onMouseDown(e);
        }

        this.setState({
            isMouseDown: true,
            mouseHasBeenDown: false,
        });
    }

    getMouseDownClasses = () => {
        const { isMouseDown, mouseHasBeenDown } = this.state;

        return {
            [styles.mouseIsDown]: isMouseDown,
            [styles.mouseHasBeenDown]: mouseHasBeenDown,
        };
    };

    getButtonVariantClasses = () => {
        const { contained, outlined, frameless } = this.props;

        return {
            [styles.contained]: contained || (!outlined && !frameless), // Default to contained
            [styles.outlined]: outlined,
            [styles.frameless]: frameless,
            [styles.icon]: false,
        };
    };

    getButtonStyleClasses = () => {
        const { primary, signal } = this.props;
        return {
            [styles.primary]: primary,
            [styles.signal]: signal,
        };
    };

    getButtonSizeClasses = displayType => {
        const { small, medium, large } = this.props;

        return {
            [styles.small]:
                small || displayType === componentDisplayTypes.compact,
            [styles.medium]:
                (medium || (!small && !large)) &&
                displayType !== componentDisplayTypes.compact, // Default to medium
            [styles.large]:
                large && displayType !== componentDisplayTypes.compact,
        };
    };

    getButtonClasses = (displayType, props) =>
        classNames(
            props.className,
            styles.container,
            {
                [styles.block]: props.block,
                [styles.flex]: props.flex,
            },
            this.getMouseDownClasses(props),
            this.getButtonVariantClasses(props),
            this.getButtonStyleClasses(props),
            this.getButtonSizeClasses(displayType)
        );

    renderButtonContent() {
        const { children } = this.props;
        return <span className={styles.button}>{children}</span>;
    }

    renderButton(displayType) {
        const {
            disabled,
            onClick,
            onClickCapture,
            url,
            targetBlank,
        } = this.props;

        const containerClassNames = this.getButtonClasses(displayType, {
            ...this.props,
            ...this.state,
        });

        if (url) {
            return (
                <a
                    className={containerClassNames}
                    disabled={disabled}
                    href={url}
                    target={targetBlank ? "_blank" : "_self"}
                    onMouseDown={() => this.onMouseDown()}
                    onMouseUp={() => this.onMouseUp()}
                    onClickCapture={onClickCapture}
                    ref={button => {
                        this.button = button;
                    }}
                >
                    {this.renderButtonContent()}
                </a>
            );
        }
        return (
            <button
                type="button"
                className={containerClassNames}
                disabled={disabled}
                onClick={onClick}
                onMouseDown={e => this.onMouseDown(e)}
                onMouseUp={() => this.onMouseUp()}
                onClickCapture={onClickCapture}
                ref={button => {
                    this.button = button;
                }}
            >
                {this.renderButtonContent()}
            </button>
        );
    }

    render() {
        return (
            <ComponentDisplayContext.Consumer>
                {displayType => this.renderButton(displayType)}
            </ComponentDisplayContext.Consumer>
        );
    }
}

export default Button;
