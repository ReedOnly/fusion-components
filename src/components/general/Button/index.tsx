import * as React from 'react';
import ComponentDisplayContext from '../../contexts/componentDisplayContext';
import ButtonComponent from './components/Button';
import AnchorComponent from './components/Anchor';

type ButtonProps = {
    disabled?: boolean;

    // Variants
    /** Use the contained variant */
    contained?: boolean;
    /** Use the outlined variant */
    outlined?: boolean;
    /** Use the frameless variant */
    frameless?: boolean;

    // Styles
    /** Use the primary coloring */
    primary?: boolean;
    /** Use the signal coloring */
    signal?: boolean;

    // Sizes
    /** Compact button */
    compact?: boolean;
    /** Comfortable button */
    comfortable?: boolean;

    // Link
    /** Provide an url and the button wil use <a>-tag instead of <button>-tag */
    url?: string | null;
    /** Set target="_blank" */
    targetBlank?: boolean;
    onMouseDown?: React.EventHandler<React.SyntheticEvent>;
    onClickCapture?: React.EventHandler<React.SyntheticEvent>;
    onClick?: React.EventHandler<React.SyntheticEvent>;
};

const Button = React.forwardRef<HTMLElement | null, React.PropsWithChildren<ButtonProps>>(
    (props, ref) => {
        // Used to apply "radar" animation on mouse up
        const [mouseHasBeenDown, setMouseHasBeenDown] = React.useState(false);
        const displayType = React.useContext<string>(ComponentDisplayContext);

        const buttonRef =
            (ref as React.RefObject<HTMLElement | null>) || React.createRef<HTMLElement | null>();

        const handleOnMouseDown = (e: React.SyntheticEvent) => {
            const { onMouseDown } = props;

            setMouseHasBeenDown(false);

            if (onMouseDown) {
                onMouseDown(e);
            }
        };

        const handleOnMouseUp = () => {
            setMouseHasBeenDown(true);
            // Focus is automatically set to the button on click,
            // but we don't want it to have focus after the click
            // action is performed. Only when tabbing to the button
            if (buttonRef.current !== null) {
                buttonRef.current.blur();
            }
        };

        return props.url ? (
            <AnchorComponent
                {...props}
                displayType={displayType}
                ref={buttonRef as React.RefObject<HTMLAnchorElement | null>}
                mouseHasBeenDown={mouseHasBeenDown}
                onMouseDown={handleOnMouseDown}
                onMouseUp={handleOnMouseUp}
                onClick={props.onClick}
                onClickCapture={props.onClickCapture}
                url={props.url}
            />
        ) : (
            <ButtonComponent
                {...props}
                displayType={displayType}
                ref={buttonRef as React.RefObject<HTMLButtonElement | null>}
                mouseHasBeenDown={mouseHasBeenDown}
                onMouseDown={handleOnMouseDown}
                onMouseUp={handleOnMouseUp}
                onClick={props.onClick}
                onClickCapture={props.onClickCapture}
            />
        );
    }
);

Button.displayName = 'Button';

Button.defaultProps = {
    disabled: false,
    contained: false,
    outlined: false,
    frameless: false,
    primary: true,
    signal: false,
    compact: false,
    comfortable: true,
    url: null,
    targetBlank: false,
    onMouseDown: () => {},
    onClickCapture: () => {},
    onClick: () => {},
};

export default Button;
