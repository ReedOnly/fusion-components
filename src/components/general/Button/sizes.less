@import './variables.less';

.button-size(@height-multiplier, @padding-multiplier) {
  padding: calc(var(--click-area) * 1px);
  margin: calc(var(--click-area) * -1px);
  height: calc((var(--grid-unit) * @height-multiplier) + (var(--click-area) * 2px));

  .button {
    font-size: 14px;
    padding: 0 calc(var(--grid-unit) * @padding-multiplier);
    height: calc(var(--grid-unit) * @height-multiplier);
    line-height: calc(var(--grid-unit) * @height-multiplier);
  }

  &.icon {
    .button {
      padding: 0 calc(var(--grid-unit) * @padding-multiplier) 0
        calc(var(--grid-unit) * @height-multiplier + var(--grid-unit) * 2);

      .icon {
        width: calc(var(--grid-unit) * @height-multiplier);
        height: calc(var(--grid-unit) * @height-multiplier);
      }
    }
  }
}

.button-sizes() {
  &.small {
    .button-size(3px, 1px);
  }

  &.medium {
    .button-size(4px, 2px);
  }

  &.large {
    .button-size(6px, 2px);
  }
}
