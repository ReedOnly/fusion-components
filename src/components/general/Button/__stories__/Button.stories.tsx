import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { actions } from '@storybook/addon-actions';
import Button from '../index';

const eventsFromNames = actions('onClick');

const createButtonStory = props => () => (
    <React.Fragment>
        <div style={{ padding: 8 }}>
            <Button primary compact {...props} {...eventsFromNames}>
                Compact button
            </Button>
            <Button primary comfortable {...props} {...eventsFromNames}>
                Comfortable button
            </Button>
            <Button primary disabled {...props} {...eventsFromNames}>
                Disabled button
            </Button>
        </div>
    </React.Fragment>
);

storiesOf('General components/Button', module)
    .addParameters({ jest: ['Button.stories.jsx'] })
    .add('Contained', createButtonStory({ contained: true }))
    .add('Outlined', createButtonStory({ outlined: true }))
    .add('Frameless', createButtonStory({ frameless: true }));
