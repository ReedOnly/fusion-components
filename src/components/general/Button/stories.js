import React from "react";
import { storiesOf } from "@storybook/react";
import { actions } from "@storybook/addon-actions";
import Button from "./";

const eventsFromNames = actions("onClick");

const createButtonStory = props => () => {
    return (
        <React.Fragment>
            <div style={{ padding: 8 }}>
                <Button primary small {...props} {...eventsFromNames}>Small button</Button>
                <Button primary {...props} {...eventsFromNames}>Medium button</Button>
                <Button primary large {...props} {...eventsFromNames}>Large button</Button>
                <Button primary disabled {...props} {...eventsFromNames}>Disabled button</Button>
            </div>
            <div style={{ padding: 8 }}>
                <Button signal small {...props} {...eventsFromNames}>Small button</Button>
                <Button signal {...props} {...eventsFromNames}>Medium button</Button>
                <Button signal large {...props} {...eventsFromNames}>Large button</Button>
                <Button signal disabled {...props} {...eventsFromNames}>Disabled button</Button>
            </div>
        </React.Fragment>
    );
};

storiesOf("General components/Button", module)
    .add("Contained", createButtonStory({ contained: true }))
    .add("Outlined", createButtonStory({ outlined: true }))
    .add("Frameless", createButtonStory({ frameless: true }));