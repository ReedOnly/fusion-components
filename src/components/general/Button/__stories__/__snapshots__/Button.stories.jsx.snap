// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General|Button Contained 1`] = `
Array [
  <div
    style={
      Object {
        "padding": 8,
      }
    }
  >
    <button
      className="container contained primary compact"
      disabled={false}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Compact button
      </span>
    </button>
    <button
      className="container contained primary comfortable"
      disabled={false}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Comfortable button
      </span>
    </button>
    <button
      className="container contained primary comfortable"
      disabled={true}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Disabled button
      </span>
    </button>
  </div>,
  <div
    style={
      Object {
        "padding": 8,
      }
    }
  >
    <button
      className="container contained primary signal compact"
      disabled={false}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Compact button
      </span>
    </button>
    <button
      className="container contained primary signal comfortable"
      disabled={false}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Comfortable button
      </span>
    </button>
    <button
      className="container contained primary signal comfortable"
      disabled={true}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Disabled button
      </span>
    </button>
  </div>,
]
`;

exports[`General|Button Frameless 1`] = `
Array [
  <div
    style={
      Object {
        "padding": 8,
      }
    }
  >
    <button
      className="container frameless primary compact"
      disabled={false}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Compact button
      </span>
    </button>
    <button
      className="container frameless primary comfortable"
      disabled={false}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Comfortable button
      </span>
    </button>
    <button
      className="container frameless primary comfortable"
      disabled={true}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Disabled button
      </span>
    </button>
  </div>,
  <div
    style={
      Object {
        "padding": 8,
      }
    }
  >
    <button
      className="container frameless primary signal compact"
      disabled={false}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Compact button
      </span>
    </button>
    <button
      className="container frameless primary signal comfortable"
      disabled={false}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Comfortable button
      </span>
    </button>
    <button
      className="container frameless primary signal comfortable"
      disabled={true}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Disabled button
      </span>
    </button>
  </div>,
]
`;

exports[`General|Button Outlined 1`] = `
Array [
  <div
    style={
      Object {
        "padding": 8,
      }
    }
  >
    <button
      className="container outlined primary compact"
      disabled={false}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Compact button
      </span>
    </button>
    <button
      className="container outlined primary comfortable"
      disabled={false}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Comfortable button
      </span>
    </button>
    <button
      className="container outlined primary comfortable"
      disabled={true}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Disabled button
      </span>
    </button>
  </div>,
  <div
    style={
      Object {
        "padding": 8,
      }
    }
  >
    <button
      className="container outlined primary signal compact"
      disabled={false}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Compact button
      </span>
    </button>
    <button
      className="container outlined primary signal comfortable"
      disabled={false}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Comfortable button
      </span>
    </button>
    <button
      className="container outlined primary signal comfortable"
      disabled={true}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Disabled button
      </span>
    </button>
  </div>,
]
`;
